Define and implement block rewards

ferencdg opened the pull request:
fixes: #246

Commit:
Amending maximum coin supply based on whitepaper

Commit:
Overrode operator 'in' method in Set

Commit:
Added attributes to ConsensusParams' ctr

ferencdg commented on source/agora/test/BlockReward.d:
```diff
@@ -0,0 +1,87 @@
+/*******************************************************************************
```
Will add 2 more tests: missing signatures and missing validators

ferencdg commented on source/agora/consensus/validation/Block.d:
```diff
@@ -148,13 +148,6 @@ public string isInvalidReason (in Block block, Engine engine, Height prev_height
     if (only_coinbase)
         return "Block: Must contain other transactions than Coinbase";
 
-    auto expected_cb_txs = getCoinbaseTX(block.txs,
```
point 5
https://github.com/bpfkorea/agora/issues/246#issuecomment-791075528

ferencdg commented on source/agora/test/Fee.d:
```diff
@@ -214,10 +214,6 @@ unittest
     {
         createAndExpectNewBlock(Height(block_idx));
     }
-
```
point 5
https://github.com/bpfkorea/agora/issues/246#issuecomment-791075528

omerfirmak replied:
Does this mean a node can be fed blocks with wrong CB txs? Do we just trust the peer when we are catching up?

ferencdg replied:
> Does this mean a node can be fed blocks with wrong CB txs? Do we just trust the peer when we are catching up?

We are still checking if at least 50% of the validators signed the block, so not just trusting only 1 peer. 

------ background

However, Mathias described an attack before
1. in 2020, lets say we have 20 validators, 2 are malicious, 8 are lazy(they don't sign blocks), and 10 are well-behaving. Those 10 well-behaving is able to create a block in 2020.
2. in 2022, 8 validators(from 2020) sell their private keys(6 previously lazy ones and 2 previously well-behaving one are the sellers).
3. now those 2 malicious validators from 2020 along with the newly bought private keys can dupe a 'catching up' a validator.

Our block reward(based on validators signing blocks) will try to minimize the number of lazy validators, so let's in 2020 we didn't have 8 lazy validators, just 2, so the block was signed by 6+10=16 validators... With block reward in mind the person in 2022 have to buy more private keys, because we will also change our catchup logic at some point to the following one:

each validator needs to accept the block with the MOST signatures at a specific height

----- end of background

so if we implement all above, then a catching up validator can only be duped, if someone can create a block with let's say 60%-70% of all signatures

ferencdg commented on source/agora/node/Ledger.d:
```diff
@@ -609,19 +744,20 @@ public class Ledger
 
         Returns:
-            List of expected Coinbase TXs
+            One or zero Coinbase TX
```
explanation
https://github.com/bpfkorea/agora/pull/1768/commits/0cd1c01abddd12ac3c63358d31470bb3d66a0120#diff-d25eeef12241e794638a748cea9cc8e7b124f1d61ccdaa962f7fb64ed3c66cb6R725

actually, even before my change we had numerous asserts that this function only returns 1 or less elements in the array

Commit:
Added non-throwing version of toString

ferencdg commented:
ready for review - squashed some commits and cleaned up the code

AndrejMitrovic commented on source/agora/common/Amount.d:
```diff
@@ -86,6 +86,13 @@ public struct Amount
     }
 
+    /// Non-throwing version of toString
+    public string toStringNT () const @safe nothrow
```
Why is this needed? Commit doesn't explain it.

ferencdg replied:
> Why is this needed? Commit doesn't explain it.

I had some 'nothrow' functions from which I had to print out the 'Amount', I will modify the comment to make it more obvious that this 'non-throwing' version should mainly be used for debugging.

Geod24 replied:
It's not used anywhere and stands out, I'd remove it.
If you need to debug you can use the `scope (failure) assert(0);` hack.

AndrejMitrovic commented on source/agora/consensus/BlockRewardCalc.d:
```diff
@@ -0,0 +1,650 @@
+       as opposed to distributing it in every block. This was originally meant
+       to reduce blockchain size.
+    2. We will delay the distribution of the block rewards, so for
+       blocks A .. B we will distribute the rewards at block B + Y.
```
What is Y here?

AndrejMitrovic commented on source/agora/consensus/BlockRewardCalc.d:
```diff
@@ -0,0 +1,650 @@
+       block signatures which is the basis of the block reward.
+    3. Each validator will be rewarded based on how many blocks they signed
+       and the reward will NOT be propotional to how much coin they staked.
+    4  The block reward is divided between the validators and the Foundation.
```
s/Foundation/Commons Budget?

AndrejMitrovic commented on source/agora/consensus/BlockRewardCalc.d:
```diff
@@ -0,0 +1,650 @@
+
+mixin AddLogger!();
+
+public class BlockRewardCalc
```
Needs `///` or ddoc is not generated.

AndrejMitrovic commented on source/agora/consensus/BlockRewardCalc.d:
```diff
@@ -0,0 +1,650 @@
+    public Amount[Point] getEachValidatorReward (in Height height, ushort[Point] validator_stakes,
+        out Amount remainder, double comply_perc, uint indent = 0) @safe const nothrow
+    {
+        auto indentStr = ' '.repeat(indent).array;
```
This will cause an allocation even if trace logging is disabled. Could perhaps turn it into a function?

ferencdg replied:
> This will cause an allocation even if trace logging is disabled. Could perhaps turn it into a function?

hmm, Ocean's format function(called by log.trace) is defined as 
```D
public void format (Args...) (Level level, cstring fmt, Args args) @safe
```
so it cannot just lazily evaluate the argument.

the only thing I can think of is 
```D
class IndentClass(int indentation)
{
    void toString (scope FormatterSink sink)
    {
        sink(' '.repeat(indent).array);
    }
}
```

and then create an instance of this class on the stack(or just declare it as struct), and then pass that instance to log.trace :) 

do you think it is worth the effort?

AndrejMitrovic replied:
Maybe it's overkill then. Let's fix it later if it's not a big deal.

Geod24 replied:
There's a simple fix: https://github.com/sociomantic-tsunami/ocean/blob/ba566c473249341633d723904a0b82ba1201a55b/src/ocean/util/log/Logger.d#L385-L391

AndrejMitrovic commented on source/agora/consensus/BlockRewardCalc.d:
```diff
@@ -0,0 +1,650 @@
+        Amount gcd_validator_reward = total_validator_reward;
+
+        ulong total_stakes;
+        foreach (ref stake; validator_stakes.byValue())
```
`const ref`?

AndrejMitrovic commented on source/agora/consensus/BlockRewardCalc.d:
```diff
@@ -0,0 +1,650 @@
+        min_validator_reward.div(total_stakes);
+
+        Amount actual_total_validator_reward;
+        foreach (pkey_stake; validator_stakes.byKeyValue())
```
Maybe also `const ref` here?

AndrejMitrovic commented on source/agora/consensus/BlockRewardCalc.d:
```diff
@@ -0,0 +1,650 @@
+        remainder.mustSub(actual_total_validator_reward);
+        foreach (amount_point_pair; each_validator_rewards.byKeyValue())
+            log.trace(indentStr ~ "{} -> {}", amount_point_pair.key, amount_point_pair.value);
+        log.trace(indentStr ~ "Remainder is: {}", remainder.toStringNT());
```
Can use comma instead of concatenation.

AndrejMitrovic commented on source/agora/consensus/BlockRewardCalc.d:
```diff
@@ -0,0 +1,650 @@
+        out Amount remainder, double comply_perc, uint indent = 0) @safe const nothrow
+    {
+        auto indentStr = ' '.repeat(indent).array;
+        log.trace(indentStr ~ "Calculating each validator reward");
```
Can use comma instead of concatenation.

AndrejMitrovic commented on source/agora/consensus/data/Params.d:
```diff
@@ -1,4 +1,4 @@
-/*******************************************************************************
+ 		 /*******************************************************************************
```
Oops

AndrejMitrovic commented on source/agora/node/Ledger.d:
```diff
@@ -356,6 +361,74 @@ public class Ledger
+        }
+        catch (Exception e) {
+            log.error("unable to read block at height: {}", height);
+            return keys;
```
Doesn't this just return `null` then?

AndrejMitrovic reviewed (commented):
Just a few comments.

I'm not sure whether using floating point is a good idea, there were reservations against using it in the past.

ferencdg commented:
> Just a few comments.
> 
> I'm not sure whether using floating point is a good idea, there were reservations against using it in the past.

Do you by any chance remember why floating point was not recommended? The only thing I could think is
1. double is kind of a waste, it takes 64 bytes
2. some cpu's use different floating point units and precision might differ -> it is not an issue in our case as I think 'double' type is standardized, so arithmetic on them should be the same on every fpu(vs 'real' type in dlang)
3. precision loss: 0.1 has a finite decimal representation, but infinete binary representation(0.00011001100110011011....) -> we don't care about this

the exponential function that I use for 'block reward reduction' returns a flotaing point value, so the only think I could do is represent the input(comply percentage) to that exponential function in decimal(I just multiply it by 10.000 let's say), so I can pass around that input parameter to different functions. However when I use it as the input in the exponential function, I still have to convert it back to floating point. (I don't think we want to create our own integer based exponential function like 2^x and then try to somehow fit the return value between 0 - 100 so it can be used as reduction percentage)

ferencdg commented:
thanks for the review @AndrejMitrovic, if anyone else has a comment, please let me know, so this PR could be merged this sprint

omerfirmak commented on source/agora/common/Amount.d:
```diff
@@ -86,6 +86,14 @@ public struct Amount
+    public string toStringNT () const @safe nothrow
+    {
+        try return this.toString();
+        catch (Exception e) {return "";}
```
Maybe this return string should indicate the exception.

ferencdg replied:
I was thinking returning null, but that is probably a bad idea: it would require the call site to do extra checks. So as you suggested I will just return some error message.

ferencdg commented:
rebased, let me know if any more changes needed

Geod24 commented on source/agora/common/Set.d:
```diff
@@ -79,6 +79,12 @@ public struct Set (T)
     }
 
+    /// Returns true if the set has the element e, otherwise returns false
+    public auto opBinary (string op)(in T e) nothrow const pure @safe @nogc if (op == "in")
```
Type ctor (`const`, `inout`, `shared`, etc...) always come first.
Also it's a template so you don't need to specify all attributes.

ferencdg replied:
not sure I am following this:
'Also it's a template so you don't need to specify all attributes.'

are those attributes stripped after the template instantiation? or are we expecting 
```D
return e in _set;
```
might throw for some instantiation?

Geod24 replied:
I mean that you can remove `@safe` because the compiler will check the body, see that it is `@safe`, and mark it `@safe` itself without you needing to do anything.

Geod24 commented on source/agora/test/Common.d:
```diff
@@ -0,0 +1,101 @@
+
+*******************************************************************************/
+
+module agora.test.Common;
```
That's what `Base` is for.

ferencdg replied:
'
it's not used anywhere and stands out, I'd remove it.
If you need to debug you can use the scope (failure) assert(0); hack.
'

it is used in BlockRewardCalc.getEachValidatorReward

ferencdg replied:
> That's what `Base` is for.

my plan was to put the higher level classes like Validator/EnrollmentManager to a new file as the module description suggests:

```
Common classes and utilities shared between the network tests.
    Classes and utilities in this module are considered higher level than
    in Base.d.
```

Base.d is already a monster with 2000+ lines of code. I can move those classes to Base.d though

hewison-chris replied:
Why not just make them `public` in one of the test modules and import them in the other test files. Otherwise `common` will become the next monster.

ferencdg replied:
I think

1. It would be hard to find that class if we just put it randomly into some of the test files from where we use it.
2. Someone already moved those classes into Base.d anyways in recent v0.x.x branch.

hewison-chris replied:
In that case can you rebase please.

Geod24 commented on source/agora/test/Common.d:
```diff
@@ -0,0 +1,101 @@
+// EnrollmentManager that might or migh not reveal preimage based on the value
+// current value of an atomic shared variable
+version (unittest)
+package class MissingPreImageEM : EnrollmentManager
```
It's a good idea to make them reusable, but those classes have a few issues. One of them being `public static shared UTXOSet utxo_set;`

ferencdg replied:
I have an other PR that tries to solve that problem:
https://github.com/bosagora/agora/pull/1817#issuecomment-804515733
any comment on that comment?

I do need to reuse those classes for block reward.

Geod24 commented on source/agora/consensus/BlockRewardCalc.d:
```diff
@@ -0,0 +1,651 @@
+
+*******************************************************************************/
+
+module agora.consensus.BlockRewardCalc;
```
```suggestion
module agora.consensus.Rewards;
```

Let's keep it simple and spellable

Geod24 commented on source/agora/consensus/BlockRewardCalc.d:
```diff
@@ -0,0 +1,651 @@
+       between the validators will also go to the Commons Budget
+
+    Copyright:
+        Copyright (c) 2021 BOS Platform Foundation Korea
```
```suggestion
        Copyright (c) 2019-2021 BOSAGORA Foundation
```

Geod24 commented on source/agora/consensus/data/Params.d:
```diff
@@ -125,6 +137,71 @@ public struct ConsensusConfig
 
     /// The amount of a penalty for slashed validators
     public Amount slash_penalty_amount = 10_000.coins;
+
+    /// The number of blocks between 2 block reward payout
+    /// The value of 3 means block reward payout happens at the block 1, 4, 7...
+    public ushort block_reward_gap = 10;
+
+    /// factor 'a' in the block reward reduction function of f(x) = b*e^(a*x) + c
+    public double block_reward_factor_a = 0.046;
+
+    /// factor 'b' in the block reward reduction function of f(x) = b*e^(a*x) + c
+    public double block_reward_factor_b = 1;
+
+    /// factor 'c' in the block reward reduction function of f(x) = b*e^(a*x) + c
+    public double block_reward_factor_c = 1;
+
+    /// The delay (measured in blocks) after which block reward is payed
+    /// The value of 5 means the block reward payout for confirming blocks
+    /// [(X-confirmation_payout_gap + 1), (X)] happens at block (X + block_reward_delay)
+    public ushort block_reward_delay = 5;
```
Do those need to be configurable ? If not they should be directly in `ConsensusParams`.

Added some information about why here: https://github.com/bosagora/agora/pull/1819

ferencdg replied:
**block_reward_factor_a
block_reward_factor_b
block_reward_factor_c**
they are very similar to already existing things like
**tx_payload_fee_factor
validator_tx_fee_cut
slash_penalty_amount**
and they are all in ConsensusConfig

**block_reward_gap
block_reward_delay**
they are very similar to already existing things like
**payout_period**
and they are all in ConsensusConfig

Maybe we want to raise a separate ticket and do all that cleanup for all variables together, so we will not have any inconsistencies.

Geod24 commented on source/agora/consensus/BlockRewardCalc.d:
```diff
@@ -0,0 +1,651 @@
+
+    1. We will only distribute block rewards in every X blocks
+       as opposed to distributing it in every block. This was originally meant
+       to reduce blockchain size.
```
Try to use a more neutral language when writing documentation. E.g. "We will only distribute" => "We will distribute". Also try to avoid usage of "we", e.g. "Block rewards are distributed every X blocks [...]".

Geod24 commented on source/agora/consensus/BlockRewardCalc.d:
```diff
@@ -0,0 +1,651 @@
+    Contains functions to calculate the block reward.
+
+    1. We will only distribute block rewards in every X blocks
+       as opposed to distributing it in every block. This was originally meant
```
originally ?

Geod24 commented on source/agora/consensus/BlockRewardCalc.d:
```diff
@@ -0,0 +1,651 @@
+       This is because we want to give enough time validators to share
+       block signatures which is the basis of the block reward.
+    3. Each validator will be rewarded based on how many blocks they signed
+       and the reward will NOT be propotional to how much coin they staked.
```
Likewise, if you feel the lack of proportionality needs emphasis, something like "[...] how many blocks they signed, regardless of the amount staked". Or "Rewards do not depend on the amount staked by a validator.".

Geod24 commented on source/agora/consensus/BlockRewardCalc.d:
```diff
@@ -0,0 +1,651 @@
+       The coins remaing is  'fixed block reward at height X' -
+       'total validator reward' will go to the Commons Budget.
+    6. The part of the total validator reward that cannot be didived equally
+       between the validators will also go to the Commons Budget
```
A few typos (`didived`, `remaing`, ...). Also this doesn't feel much like documentation, as it lacks structure.
I would recommend to state the goals first (study supply, reward active participants, etc..), then expand on the actual system property and the details of the implementation after.

Geod24 commented on source/agora/common/Amount.d:
```diff
@@ -33,7 +33,7 @@ public struct Amount
     public static immutable Amount UnitPerCoin   = Amount(10_000_000, true);
     /// Maximum number of BOA coins that can ever be in circulation
-    public static immutable ulong  MaxCoinSupply = 500_000_000;
+    public static immutable ulong  MaxCoinSupply = 4_948_159_353;
```
When `Amount` was created, the goal was simply to ensure we never go over the limit. The idea was to later add a function which would give us the coin supply based on the block height. Do you think that's something we can achieve or it would create too much trouble (e.g. by forcing us to pass extra parameters everywhere) ?

ferencdg replied:
> When `Amount` was created, the goal was simply to ensure we never go over the limit. The idea was to later add a function which would give us the coin supply based on the block height. Do you think that's something we can achieve or it would create too much trouble (e.g. by forcing us to pass extra parameters everywhere) ?

We would have problems when we want to add let's say 
```D
Amount a = Amount(10000, 2);
Amount b = Amount(20000, 4) 
```
where the 2 and 4 are the heights. Addition will no longer be associative, because 
```D
a.add(b) // might overflow(unless we change a's hight to 4)
b.add(a) // will not overflow(provided that at height 4 the amount of 30000 is okay)
```

Additionally there are around 800+ mentions of Amount(case sensitive/separate word) in our codebase, I think this could be a ticket on its own.

Geod24 commented on source/agora/consensus/BlockRewardCalc.d:
```diff
@@ -0,0 +1,651 @@
+    Returns: which 'period' the particular height falls into and whether that height
+    is the last height for that particular period
+
+    ***************************************************************************/
```
Indentation is wrong

Geod24 commented on source/agora/consensus/BlockRewardCalc.d:
```diff
@@ -0,0 +1,651 @@
+
+    ***************************************************************************/
+
+    private CurrentPeriodTup getCurrentPeriod (in Height height, immutable(BlockRewardsTup)[] block_rewards) @safe const pure nothrow
+    {
+        ulong blocks_accumulated = 0;
+        foreach (period; 0 .. block_rewards.length)
+        {
+            blocks_accumulated += block_rewards[period].dist_period_secs / consensus_params.BlockInterval.total!"seconds";
+            if (blocks_accumulated >= height)
+                return CurrentPeriodTup(period, blocks_accumulated == height);
+        }
+        return CurrentPeriodTup(block_rewards.length, false);
+    }
```
Doesn't look like it's using anything from the class, should it be either a free function, or at least a `static` one ?

ferencdg replied:
it uses consensus_params

Geod24 commented on source/agora/consensus/BlockRewardCalc.d:
```diff
@@ -0,0 +1,651 @@
+
+    ***************************************************************************/
+
+    public Amount[Point] getEachValidatorReward (in Height height, ushort[Point] validator_stakes,
+        out Amount remainder, double comply_perc, uint indent = 0) @safe const nothrow
```
Using `PublicKey` instead of `Point`

ferencdg replied:
> Using `PublicKey` instead of `Point`

I am depending on other functions like getValidatorAtIndex() that was written long time ago and still uses Point. I think it would be better to clean up all usages at once instead of me doing from Point and to Point conversation whenever I was to call a legacy function.

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -356,6 +361,74 @@ public class Ledger
+        foreach (i; 0 .. validator_cnt)
+        {
+            auto validator_pkey = enroll_man.getValidatorAtIndex(height, i);
+            if (block.header.validators[i] && !block.header.missing_validators.canFind(i))
```
The second check will never be false, because a validator that signed had to sign using its pre-image, hence there's no way to be missing yet sign. If you are missing, you get slashed.

ferencdg replied:
I can change, but we have a slashing policy, so what happens if in the future the slashing policy will not be based on pre-images or the signature will not be based pre-images?

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -356,6 +361,74 @@ public class Ledger
         return true;
     }
 
+    /***************************************************************************
+
+        Returns the public keys of the validators that signed the block at
+        the specified height, and and also not classified as 'missing' validator
+
+        Params:
+            height = block height
+
+        Returns: the public keys of the validators that signed the block at
+                 the specified height, and and also not
+                 classified as 'missing' validator
+
+    ***************************************************************************/
+
+    private Point[] getPKeysOfSignedNonMissingValidators (Height height) nothrow @safe
```
It should not be possible to be both `missing` and signing, as explained under this comment.
Also, perhaps this function could use with a better name, such as `getRewardEligibleValidators` ? It's a more semantic / abstract name, which makes sense *unless* we have a need to get the list of signers for other purposes ?

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -356,6 +361,74 @@ public class Ledger
+
+        foreach (height; start_height .. end_height + 1)
+            foreach (point; getPKeysOfSignedNonMissingValidators(height))
+                signed_cnt_keys_map[point]++;
```
Does this work ? I would expect it to throw.

ferencdg replied:
if you mean line 427, it works

```D
    int[string] m;
    m["hello"]++;
    writeln(m["hello"]);
```

writes out 1

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -356,6 +361,74 @@ public class Ledger
+    {
+        Point[] keys;
+        Block block;
+        try
+        {
+            block = this.storage.readBlock(height);
+        }
+        catch (Exception e) {
+            log.error("unable to read block at height: {}", height);
+            return null;
+        }
```
This is not handled at the call site. In general, we should probably just let it `throw`, much better to do so than to propose invalid consensus data.

ferencdg replied:
This function is very deep in the call hierarchy, and its parents are all nothrow. I can change it, but that will remove nothrow from core functions like valueExternalized, getValidTXSet.

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -579,17 +656,77 @@ public class Ledger
+            double total_actual_sigcount = pkey_sigcounts.byValue().fold!((a, b) => a + b)(0);
+            log.trace("Total actual signatures count: {}", total_actual_sigcount);
+            double total_expected_sigcount = iota(period_start.value, period_end.value + 1)
+                .map!(height => this.enroll_man.getCountOfValidators(Height(height))).fold!((a, b) => a + b)(cast(size_t) 0);
```
Uh, why the `cast` ?

ferencdg replied:
doesn't compile without the cast

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -579,17 +656,77 @@ public class Ledger
+
+            // calculate the penalty for missing singatures
+            auto pkey_sigcounts = getPKeysOfSignedNonMissingValidators(period_start, period_end);
+            double total_actual_sigcount = pkey_sigcounts.byValue().fold!((a, b) => a + b)(0);
```
Shouldn't this be an integer ? Why double ? Also I think you can use `sum` here ;)

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -579,17 +656,77 @@ public class Ledger
+            auto pkey_sigcounts = getPKeysOfSignedNonMissingValidators(period_start, period_end);
+            double total_actual_sigcount = pkey_sigcounts.byValue().fold!((a, b) => a + b)(0);
+            log.trace("Total actual signatures count: {}", total_actual_sigcount);
+            double total_expected_sigcount = iota(period_start.value, period_end.value + 1)
+                .map!(height => this.enroll_man.getCountOfValidators(Height(height))).fold!((a, b) => a + b)(cast(size_t) 0);
```
Likewise, shouldn't this be an integer, not a double ?

Geod24 commented on source/agora/node/Ledger.d:
```diff
@@ -579,17 +656,77 @@ public class Ledger
+                if (pkey_reward.value != Amount(0))
+                {
+                    if (auto pkey_ind_found = (pkey_reward.key in pkey_to_output_ind))
+                        coinbase_tx.outputs[*pkey_ind_found].value.mustAdd(pkey_reward.value);
```
`mustAdd` shouldn't be used outside of tests

ferencdg replied:
> mustAdd

so if I use 'add', should I check if the resulting Amount is still valid? In case your answer is 'yes I should check' and it is an invalid amount after the addition, then I would like to assert(0). I am confused

Geod24 commented on source/agora/consensus/BlockRewardCalc.d:
```diff
@@ -0,0 +1,651 @@
+
+    ***************************************************************************/
+
+    public bool is_payout_time (Height height) @safe const pure nothrow
```
`isPayoutTime` according to our style guide

Geod24 commented on source/agora/consensus/BlockRewardCalc.d:
```diff
@@ -0,0 +1,651 @@
+        this.consensus_params = consensus_params;
+    }
+
+    ///
+    private alias CurrentPeriodTup = Tuple!(ulong, "period", bool, "is_period_end");
```
A `struct` would also work :)

ferencdg replied:
Just wondering if struct would have any benefit over Tuples or if we should refine our coding guidelines. Tuples have some nice methods like expand, translate... Any suggestion on when to use one over the other?

Geod24 commented on source/agora/consensus/BlockRewardCalc.d:
```diff
@@ -0,0 +1,651 @@
+        min_validator_reward.div(total_stakes);
+
+        Amount actual_total_validator_reward;
+        foreach (const ref pkey_stake; validator_stakes.byKeyValue())
```
Why the complicated pattern when you can just do:
```D
foreach (key; const ref value; validator_stakes)
```

?

ferencdg replied:
got a nasty DLang bug when tried:
source/agora/consensus/BlockRewardCalc.d(146,9): Error: _aaApply2 is not nothrow
source/agora/consensus/BlockRewardCalc.d(128,26): Error: nothrow function agora.consensus.BlockRewardCalc.BlockRewardCalc.getEachValidatorReward may throw

to reproduce
```D
void f() nothrow
{
    int[string] m;
    foreach(const ref k, v; m)
    {
    }
}
int testmain()
{
    f();
    return 0;
}
```

Geod24 commented on source/agora/consensus/BlockRewardCalc.d:
```diff
@@ -0,0 +1,651 @@
+        Amount remainder;
+        auto block_reward = getBlockRewardCalc(2, 3, cast(immutable(BlockRewardsTup)[])[BlockRewardsTup(Amount(100), 10)], 1.seconds);
+
+        auto validator1 = Point.fromString("0xab4f6f6e85b8d0d38f5d5798a4bdc4dd444c8909c8a5389d3bb209a186105110");
+        auto validator2 = Point.fromString("0xab4f6f6e85b8d0d38f5d5798a4bdc4dd444c8909c8a5389d3bb209a186105111");
```
Use well-known KPs

Geod24 commented on source/agora/consensus/BlockRewardCalc.d:
```diff
@@ -0,0 +1,651 @@
+
+    ***************************************************************************/
+
+    public Amount getReducedValidatorReward (in Height height, double comply_perc, uint indent = 0) @safe const nothrow
```
Let me tell you a floating point joke:
A man has a computer problem. He decides to use floating point. Now he has 2.00000000001 problems.
Usually, we use an integer with a fixed precision for this kind of use case.

ferencdg replied:
ehhmm, I tried to explain it here:
https://github.com/bosagora/agora/pull/1768#issuecomment-799178954

any comment on that?

Geod24 commented on source/agora/consensus/BlockRewardCalc.d:
```diff
@@ -0,0 +1,651 @@
+
+        if (reduce_perc == 100)
+            return total_validator_reward;
+        total_validator_reward.div(100);
```
What about the remainder ?

ferencdg replied:
remainder goes to the Commons Budget and calculated by: getExtraCommonsBudgetReward

AndrejMitrovic commented on source/agora/consensus/BlockRewardCalc.d:
```diff
@@ -0,0 +1,651 @@
+       This is because we want to give enough time validators to share
+       block signatures which is the basis of the block reward.
+    3. Each validator will be rewarded based on how many blocks they signed
+       and the reward will NOT be propotional to how much coin they staked.
```
I don't quite clearly see why it shouldn't be proportional? Is this explained anywhere?

ferencdg replied:
If I remember correctly it was not my decision, but let me know if I am mistaken (@Geod24). A validator who has X * 40k BOA can and will run multiple Agora instances to maximimze its reward, so I agree with you on this.

Geod24 reviewed (commented):
Did a first pass.

Commit:
Added 'enabled' method to Log.d

current opDispatch implementaion failes, because
1. return value is void
2. attributes like nothrow are missing

ferencdg commented:
thanks for the review @Geod24, I fixed what I could, but for the following ones I had suggestions/questions:

https://github.com/bosagora/agora/pull/1768#discussion_r600189742
https://github.com/bosagora/agora/pull/1768#discussion_r600213266
https://github.com/bosagora/agora/pull/1768#discussion_r601000883
https://github.com/bosagora/agora/pull/1768#discussion_r600199453
https://github.com/bosagora/agora/pull/1768#discussion_r601009640
https://github.com/bosagora/agora/pull/1768#discussion_r600201037

If I could get some comment on those that you disagree, then I might be able to finish this task this Sprint.

ferencdg commented:
rebase #3

Commit:
Added block reward

fixes: #246

ferencdg commented:
rebase #4

hewison-chris commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -241,6 +244,27 @@ public class EnrollmentManager
+        else
+            return this.keymap.index_to_key[height][index];
+    }
+
```
Would prefer this in a separate commit (or with the changes it is created for) as it is unrelated to the other changes in this commit.

hewison-chris reviewed (commented):
I will continue the review after reading the `Rewards` section of the white paper.

ferencdg commented on source/agora/node/Ledger.d:
```diff
@@ -787,11 +929,6 @@ public class Ledger
             return "Internal error: Could not calculate random seed at current height";
         }
 
-        auto expected_cb_txs = this.getCoinbaseTX(block.txs, block.header.missing_validators);
```
https://github.com/bosagora/agora/pull/1768#discussion_r592191394

ferencdg commented:
rebase #5

hewison-chris commented:
Closing as will continue in new PR https://github.com/bosagora/agora/pull/2415
